File: trunk/src/scala/org/codehaus/testdox/intellij/TemplateNameResolver.scala
package org.codehaus.testdox.intellij

import java.beans.PropertyChangeEvent
import java.beans.PropertyChangeListener

import org.codehaus.testdox.intellij.TemplateNameResolver._
import org.codehaus.testdox.intellij.config.Configuration

class TemplateNameResolver(configuration: Configuration) extends NameResolver with PropertyChangeListener {

  private var prefix = ""
  private var suffix = ""

  configureTemplate(configuration.getTestNameTemplate)
  configuration.addPropertyChangeListener(this)

  def propertyChange(event: PropertyChangeEvent) {
    if (Configuration.TEST_NAME_TEMPLATE == event.getPropertyName()) {
      configureTemplate(event.getNewValue().asInstanceOf[String])
    }
  }

  private def configureTemplate(template: String) {
    prefix = ""
    suffix = ""

    val tokens = template.split(NAME_TOKEN)
    if (tokens.length > 0) {
      if (tokens.length >= 2) {
        prefix = tokens(0)
        suffix = tokens(1)
      } else if (template.startsWith(NAME_TOKEN)) {
        suffix = tokens(0)
      } else if (template.endsWith(NAME_TOKEN)) {
        prefix = tokens(0)
      }
    }
  }

  def getRealClassName(className: String): String = {
    if (isRealClass(className)) return className

    var classNameOnly = extractClassNameWithoutPackageName(className)
    if (classNameOnly.startsWith(prefix)) {
      classNameOnly = classNameOnly.substring(prefix.length())
    }
    if (classNameOnly.endsWith(suffix)) {
      classNameOnly = classNameOnly.substring(0, classNameOnly.length() - suffix.length())
    }
    extractPackagePrefix(className) + classNameOnly
  }

  def getTestClassName(className: String): String = {
    if (isTestClass(className)) return className 

    val packagePrefix = extractPackagePrefix(className)
    val classNameOnly = extractClassNameWithoutPackageName(className)
    packagePrefix + prefix + classNameOnly + suffix
  }

  def getRealClassNameForDisplay(className: String) = {
    if (configuration.getShowFullyQualifiedClassName)
      getRealClassName(className)
    else
      extractClassNameWithoutPackageName(getRealClassName(className))
  }

  def isTestClass(className: String) = className != null && !isRealClass(className)

  def isRealClass(className: String): Boolean = {
    if (className == null) return false

    val classNameOnly = extractClassNameWithoutPackageName(className)
    !(classNameOnly.startsWith(prefix) && classNameOnly.endsWith(suffix))
  }

  private def isClassUnderPackage(className: String) = className.contains('.')

  private def extractClassNameWithoutPackageName(className: String) = {
    if (isClassUnderPackage(className))
      className.substring(className.lastIndexOf('.') + 1)
    else
      className
  }

  private def extractPackagePrefix(className: String) = {
    if (isClassUnderPackage(className))
      className.substring(0, className.lastIndexOf('.') + 1)
    else
      ""
  }
}

object TemplateNameResolver {
  val NAME_TOKEN = "<classname>"
  val DEFAULT_TEMPLATE = NAME_TOKEN + "Test"
  val DEFAULT_PREFIX = "test"
}

File: branches/pre-scala-migration/src/java/org/codehaus/testdox/intellij/TemplateNameResolver.java
package org.codehaus.testdox.intellij;

import java.beans.PropertyChangeEvent;
import java.beans.PropertyChangeListener;

import org.codehaus.testdox.intellij.config.ConfigurationBean;

public class TemplateNameResolver implements NameResolver, PropertyChangeListener {

    public static final String NAME_TOKEN = "<classname>";
    public static final String DEFAULT_TEMPLATE = NAME_TOKEN + "Test";
    public static final String DEFAULT_PREFIX = "test";

    private final ConfigurationBean configuration;

    private String prefix;
    private String suffix;

    public TemplateNameResolver(ConfigurationBean configuration) {
        this.configuration = configuration;

        configureTemplate(configuration.getTestNameTemplate());
        configuration.addPropertyChangeListener(this);
    }

    private void configureTemplate(String template) {
        prefix = "";
        suffix = "";
        String[] tokens = template.split(NAME_TOKEN);
        if (tokens.length > 0) {
            if (tokens.length >= 2) {
                prefix = tokens[0];
                suffix = tokens[1];
            } else if (template.startsWith(NAME_TOKEN)) {
                suffix = tokens[0];
            } else if (template.endsWith(NAME_TOKEN)) {
                prefix = tokens[0];
            }
        }
    }

    public void propertyChange(PropertyChangeEvent evt) {
        if (ConfigurationBean.TEST_NAME_TEMPLATE.equals(evt.getPropertyName())) {
            configureTemplate((String) evt.getNewValue());
        }
    }

    public String getRealClassName(String className) {
        if (isRealClass(className)) {
            return className;
        }
        String classNameOnly = extractClassNameWithoutPackageName(className);
        if (classNameOnly.startsWith(prefix)) {
            classNameOnly = classNameOnly.substring(prefix.length());
        }
        if (classNameOnly.endsWith(suffix)) {
            classNameOnly = classNameOnly.substring(0, classNameOnly.length() - suffix.length());
        }
        return extractPackagePrefix(className) + classNameOnly;
    }

    public String getTestClassName(String className) {
        if (isTestClass(className)) {
            return className;
        }

        String packagePrefix = extractPackagePrefix(className);
        String classNameOnly = extractClassNameWithoutPackageName(className);
        return packagePrefix + prefix + classNameOnly + suffix;
    }

    public String getRealClassNameForDisplay(String className) {
        return (configuration.isShowFullyQualifiedClassName())
                ? getRealClassName(className)
                : extractClassNameWithoutPackageName(getRealClassName(className));
    }

    private boolean isClassUnderPackage(String className) {
        return (className.indexOf('.') > -1);
    }

    private String extractPackagePrefix(String className) {
        if (isClassUnderPackage(className)) {
            return className.substring(0, className.lastIndexOf('.') + 1);
        }
        return "";
    }

    public String extractClassNameWithoutPackageName(String className) {
        if (isClassUnderPackage(className)) {
            return className.substring(className.lastIndexOf('.') + 1);
        }
        return className;
    }

    public boolean isRealClass(String className) {
        if (className == null) {
            return false;
        }
        String classNameOnly = extractClassNameWithoutPackageName(className);
        return !(classNameOnly.startsWith(prefix) && classNameOnly.endsWith(suffix));
    }

    public boolean isTestClass(String className) {
        if (className == null) {
            return false;
        }
        return !isRealClass(className);
    }
}
